File: php/Fapshi.php
<?php 
class Fapshi {

    const BASEURL = 'https://live.fapshi.com';
    const HEADERS = array(
        'apiuser: replace_me_with_apiuser',
        'apikey: replace_me_with_apikey',
        'Content-Type: application/json'
    );
    const ERRORS = array(
        'invalid type, string expected',
        'invalid type, array expected',
        'amount required',
        'amount must be of type integer',
        'amount cannot be less than 100 XAF',
    );


    public function initiate_pay(array $data) : array {
        if(!is_array($data)){
            $error = array('message'=>Fapshi::ERRORS[1],'statusCode'=>400);
        }
        else if(!array_key_exists('amount', $data)){
            $error = array('message'=>Fapshi::ERRORS[2],'statusCode'=>400);
        }
        else if(!is_int($data['amount'])){
            $error = array('message'=>Fapshi::ERRORS[3],'statusCode'=>400);
        }
        else if($data['amount']<100){
            $error = array('message'=>Fapshi::ERRORS[4],'statusCode'=>400);
        }
        if(isset($error)){
            return $error;
        }

        $url = Fapshi::BASEURL.'/initiate-pay';
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => json_encode($data),
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        $response['statusCode'] = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        return $response;
    }
    

    public function direct_pay(array $data) : array {
        if(!is_array($data)){
            $error = array('message'=>Fapshi::ERRORS[0],'statusCode'=>400);
        }
        else if(!array_key_exists('amount', $data)){
            $error = array('message'=>Fapshi::ERRORS[2],'statusCode'=>400);
        }
        else if(!is_int($data['amount'])){
            $error = array('message'=>Fapshi::ERRORS[3],'statusCode'=>400);
        }
        else if($data['amount']<100){
            $error = array('message'=>Fapshi::ERRORS[4],'statusCode'=>400);
        }
        else if(!array_key_exists('phone', $data)){
            $error = array('message'=>'phone number required','statusCode'=>400);
        }
        else if(!is_string($data['phone'])){
            $error = array('message'=>'phone must be of type string','statusCode'=>400);
        }
        else if(!preg_match('/^6[0-9]{8}$/', $data['phone'])){
            $error = array('message'=>'invalid phone number','statusCode'=>400);
        }
        if(isset($error)){
            return $error;
        }

        $url = Fapshi::BASEURL.'/direct-pay';
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => json_encode($data),
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        $response['statusCode'] = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        return $response;
    }


    public function payment_status(string $transId) : array {
        if(!is_string($transId) || empty($transId)){
            return array('message'=>Fapshi::ERRORS[0],'statusCode'=>400);
        }
        if(!preg_match('/^[a-zA-Z0-9]{8,10}$/', $transId)){
            return array('message'=>'invalid transaction id','statusCode'=>400);
        }

        $url = Fapshi::BASEURL.'/payment-status/'.$transId;
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'GET',
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        $response['statusCode'] = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        return $response;
    }


    public function expire_pay(string $transId) : array {
        if(!is_string($transId) || empty($transId)){
            return array('message'=>Fapshi::ERRORS[0],'statusCode'=>400);
        }
        if(!preg_match('/^[a-zA-Z0-9]{8,10}$/', $transId)){
            return array('message'=>'invalid transaction id','statusCode'=>400);
        }

        $data = array('transId'=> $transId);
        $url = Fapshi::BASEURL.'/expire-pay';
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => json_encode($data),
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        $response['statusCode'] = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        return $response;
    }


    public function get_user_trans(string $userId) : array {
        if(!is_string($userId) || empty($userId)){
            return array('message'=>Fapshi::ERRORS[0],'statusCode'=>400);
        }
        if(!preg_match('/^[a-zA-Z0-9-_]{1,100}$/', $userId)){
            return array('message'=>'invalid user id','statusCode'=>400);
        }

        $url = Fapshi::BASEURL.'/transaction/'.$userId;
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'GET',
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        curl_close($curl);
        return $response;
    }

    
    public function balance() : array {
        $url = Fapshi::BASEURL.'/balance';
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'GET',
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        $response['statusCode'] = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        return $response;
    }


    public function payout(array $data) : array {
        if(!is_array($data)){
            $error = array('message'=>Fapshi::ERRORS[0],'statusCode'=>400);
        }
        else if(!array_key_exists('amount', $data)){
            $error = array('message'=>Fapshi::ERRORS[2],'statusCode'=>400);
        }
        else if(!is_int($data['amount'])){
            $error = array('message'=>Fapshi::ERRORS[3],'statusCode'=>400);
        }
        else if($data['amount']<100){
            $error = array('message'=>Fapshi::ERRORS[4],'statusCode'=>400);
        }
        else if(!array_key_exists('phone', $data)){
            $error = array('message'=>'phone number required','statusCode'=>400);
        }
        else if(!is_string($data['phone'])){
            $error = array('message'=>'phone must be of type string','statusCode'=>400);
        }
        else if(!preg_match('/^6[0-9]{8}$/', $data['phone'])){
            $error = array('message'=>'invalid phone number','statusCode'=>400);
        }
        if(isset($error)){
            return $error;
        }

        $url = Fapshi::BASEURL.'/payout';
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => json_encode($data),
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        $response['statusCode'] = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        return $response;
    }


    public function search(array $params) : array {

        $url = Fapshi::BASEURL.'/search?'.http_build_query($params);
        $curl = curl_init();

        curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => 'GET',
            CURLOPT_HTTPHEADER => Fapshi::HEADERS,
        ));

        $response = curl_exec($curl);
        $response = json_decode($response, true);
        curl_close($curl);
        return $response;
    }

}

File: golang/fapshi.go
package fapshi

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
)

const (
	BaseURL = "https://live.fapshi.com/"
	Apikey  = "replace_me_with_apikey"
	ApiUser = "replace_me_with_apiuser"
)

var client = &http.Client{}

type FapshiPayment struct {
	Amount      int64  `json:"amount"`
	Phone       string `json:"phone"`
	Medium      string `json:"medium"`
	Name        string `json:"name"`
	Email       string `json:"email"`
	UserId      string `json:"userId"`
	ExternalId  string `json:"externalId"`
	RedirectUrl string `json:"redirectUrl"`
	Message     string `json:"message"`
}

func request(method, path string, data interface{}) (any, error) {
	url := BaseURL + path
	jsonData, err := json.Marshal(data)
	if err != nil {
		fmt.Print(err)
		return nil, err
	}
	req, err := http.NewRequest(method, url, bytes.NewBuffer(jsonData))
	if err != nil {
		fmt.Print(err.Error())
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json; charset=utf-8")
	req.Header.Add("apikey", Apikey)
	req.Header.Add("apiuser", ApiUser)

	res, err := client.Do(req)
	if err != nil {
		fmt.Print(err.Error())
		return nil, err
	}
	defer res.Body.Close()
	body, readErr := io.ReadAll(res.Body)
	if readErr != nil {
		return nil, readErr
	}
	fmt.Println(string(body))
	var resp interface{}
	json.Unmarshal(body, &resp)
	return resp, nil
}

/*
This function returns an object with the link were a user is to be redirected in order to complete his payment

Below is a parameter template. Just amount is required

	data = {
	    "amount": Integer ,
	    "email": String,
	    "userId": String,
	    "externalId": String,
	    "redirectUrl": String,
	    "message": String
	}
*/

func InitiatePay(data FapshiPayment) (any, error) {
	if data.Amount < 100 {
		return nil, fmt.Errorf("Amount should be greater than or equal to 100")
	}
	resp, err := request(http.MethodPost, "initiate-pay", data)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

/*
This function directly initiates a payment request to a user's mobile device and
returns an object with a transId property that is used to get the status of the payment

Below is a parameter template. amount and phone are required

	data = {
	    "amount": Integer ,
	    "phone": String ,
	    "medium": String,
	    "name": String,
	    "email": String,
	    "userId": String,
	    "externalId": String,
	    "message": String
	}
*/
func DirectPay(data FapshiPayment) (any, error) {
	if data.Amount < 100 {
		return nil, fmt.Errorf("Amount should be greater than or equal to 100")
	}
	if data.Phone == "" {
		return nil, fmt.Errorf("Phone number is required")
	}
	match, err := regexp.MatchString(`^6[\d]{8}$`, data.Phone)
	if err != nil {
		return nil, err
	}
	if !match {
		return nil, fmt.Errorf("Invalid phone number")
	}
	resp, err := request(http.MethodPost, "direct-pay", data)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

/*
This function returns an object containing the details of the transaction associated with the Id passed as parameter
*/
func PaymentStatus(transId string) (any, error) {
	if transId == "" {
		return nil, fmt.Errorf("Transaction Id is required")
	}
	match, err := regexp.MatchString(`^[a-zA-Z0-9]{1,100}$`, transId)
	if err != nil {
		return nil, err

	}
	if !match {
		return nil, fmt.Errorf("Invalid Transaction Id")
	}
	path := "payment-status/" + transId
	resp, err := request(http.MethodGet, path, nil)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

/*
This function expires the transaction associated with the Id passed as parameter and returns an object containing the details of the transaction
*/
func ExpirePay(transId string) (any, error) {
	if transId == "" {
		return nil, fmt.Errorf("Transaction Id is required")
	}
	match, err := regexp.MatchString(`^[a-zA-Z0-9]{1,100}$`, transId)
	if err != nil {
		return nil, err
	}
	if !match {
		return nil, fmt.Errorf("Invalid Transaction Id")
	}
	data := map[string]string{
		"transId": transId,
	}
	resp, err := request(http.MethodPost, "expire-pay", data)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

/*
This function returns an array of objects containing the transaction details of the user Id passed as parameter
*/
func UserTrans(userId string) (any, error) {
	if userId == "" {
		return nil, fmt.Errorf("User Id is required")
	}
	match, err := regexp.MatchString(`^[a-zA-Z0-9]{1,100}$`, userId)
	if err != nil {
		return nil, err
	}
	if !match {
		return nil, fmt.Errorf("Invalid User Id")
	}
	path := "transactions/" + userId
	resp, err := request(http.MethodGet, path, nil)
	if err != nil {
		return nil, err
	}
	return resp, nil
}
